airbnb_project/listings/harvester_app/harvester/pipelines.py

```diff
@@ -102,3 +105,80 @@ def _extract_bathrooms(self, adapter):
             print(f"Error extracting bathrooms: {e}")
 
         adapter['baths'] = bathroom
+
+
+class DjangoORMPipeline:
+    """
+       A Django ORM pipeline for processing and storing Airbnb listing data.
+
+       This pipeline handles the creation of new Listing objects in the database,
+       ensuring no duplicate listings are created based on the airbnb_listing_id.
+       All fields are required and must be present in the input item. The pipeline
+       includes error handling for database integrity issues and logging capabilities
+       for monitoring the data processing flow.
+    """
+    def process_item(self, item, spider):
+        """
+        Process a scraped item and store it in the database if valid and unique.
+
+        Args:
+            item (dict): A dictionary containing scraped Airbnb listing data with these required keys:
+                - airbnb_listing_id (str): Unique identifier for the Airbnb listing
+                - name (str): Name of the listing
+                - title (str): Title of the listing
+                - baths (str): Number of bathrooms
+                - beds (str): Number of beds
+                - latitude (str): Latitude coordinate
+                - longitude (str): Longitude coordinate
+                - person_capacity (str): Maximum number of guests
+                - registration_number (str): Official registration number
+                - room_type (str): Type of room/accommodation
+                - location (str): Location description
+                - bath_is_shared (bool): Whether bathroom is shared
+                - baths_text (str): Textual description of bathroom facilities
+            spider: The spider instance that is running the crawl
+
+        Returns:
+            dict: The original item dictionary, unmodified
+
+        Note:
+            - All fields are required - missing or empty airbnb_listing_id will cause the item to be logged and skipped
+            - If a listing with the same airbnb_listing_id exists, it will be skipped
+        """
+        airbnb_listing_id = item.get('airbnb_listing_id', '')
+
+        # Early return if no valid airbnb_listing_id
+        if not airbnb_listing_id:
+            spider.logger.error("Missing required airbnb_listing_id, skipping item")
```

Comment:
do we want to log any other item identifier?

Reply:
Yes, I think the url of the page the listing is scarped from could be useful. I will see what else that is useful can be logged here.

Reply:
you can just `json.dumps` the entire item.